pkg/migrator/migrator.go

```diff
@@ -0,0 +1,231 @@
+			datastoreCs, err = findCsFromOpsrc(&instance, kubeClient)
+			if err != nil {
+				logger.Errorf("[migration] Could not infer datastore CatalogSource for Subscription %s.", instance.GetName())
+				continue
+			}
+		}
+		// point the subscription to the datastore CatalogSource
+		updatedSub := setSubsCatalogSource(&instance, datastoreCs.GetName(), datastoreCs.GetNamespace())
+		err = kubeClient.Update(context.TODO(), updatedSub)
+		if err != nil {
+			logger.Errorf("[migration] Error updating subscription %s. Error: %s", instance.GetName(), err.Error())
+		} else {
+			logger.Println("[migration] Succesfully updated Subscription %s", instance.GetName())
+		}
+	}
+	return installedCscs, nil
+}
+
+// deleteInstalledCscs deletes the CSCs installed during opeartor installation.
+// The child resources of the CSCs are delete by the finalizer.
+func deleteInstalledCscs(cscs []types.NamespacedName, kubeClient client.Client, logger *logrus.Entry) {
+	for _, cscInfo := range cscs {
+		csc := helperNewCatalogSourceConfig(cscInfo.Namespace, cscInfo.Name)
+		err := kubeClient.Delete(context.TODO(), csc)
+		if err != nil {
+			logger.Errorf("[migration] Failed to delete installed CSC %s with error: ", cscInfo.Name, err.Error())
+		} else {
+			logger.Printf("[migration] Stale CSC %s scheduled for deletion.", cscInfo.Name)
+		}
+	}
+}
+
+// deleteDatastoreCscs deletes the datastore CSCs created by OperatorSources.
+// The child resources of the CSCs are delete by the finalizer.
+func deleteDatastoreCscs(kubeClient client.Client, logger *logrus.Entry) error {
+	options := &client.ListOptions{}
+	options.SetLabelSelector(fmt.Sprintf("opsrc-datastore: \"true\""))
+
+	cscs := &v2.CatalogSourceConfigList{}
+	// Get the list of existing cscs that have the lable "opsrc-datastore: "true" "
+	err := kubeClient.List(context.TODO(), options, cscs)
+	if err != nil {
+		return err
+	}
+	for _, csc := range cscs.Items {
+		err = kubeClient.Delete(context.TODO(), &csc)
+		if err != nil {
+			logger.Errorf("[migration] Failed to delete CatalogSourceConfig %s. Error: %s", csc.GetName(), err.Error())
+		} else {
+			logger.Printf("[migration] Datastore CSC %s scheduled for deletion.", csc.GetName())
+		}
+	}
+	return nil
+}
+
+// findCatalogSource infers the datastore CatalogSource that a given existing Subscription will
+// contain information about in it's spec.source and spec.sourceNamespace fields
+func findCatalogSource(subscription *olm.Subscription, client client.Client) (*olm.CatalogSource, error) {
+	associatedCscName := subscription.GetLabels()[builders.OwnerNameLabel]
+	possibleCsName := extractCsName(associatedCscName)
+	possibleCsNamespace := subscription.GetLabels()[builders.OwnerNamespaceLabel]
+	// try and fetch the CatalogSource
+	datastoreCs := &olm.CatalogSource{}
+	namespacedName := types.NamespacedName{Name: possibleCsName, Namespace: possibleCsNamespace}
+	err := client.Get(context.TODO(), namespacedName, datastoreCs)
```

Comment:
@aravindhp trying to fetch the `CatalogSource` is more for creating a back up plan in case inferring the `CatalogSource` name from the `csc-owner-name` label fails.  
If the inferred `CatalogSource` does not exist, the `migrator` then moves onto the next strategy to find the corresponding `datastore` `CatalogSource`.

Reply:
Sure but the answer in both cases should be the same. More over if you attempted the get when someone deleted the datastore `CatalogSource` and the watch which I am adding brings it back, it will still result in a `Subscription` not being updated.

Reply:
@aravindhp there are two reasons for checking if the `CatalogSource` exists: 

A) If the datastore `CatalogSource` has been deleted then it shouldn't be referenced from the `Subscription`

Sure, with the watch you're adding it'll be brought back, but there's another scenario where:

B) The infered `CatalogSource` name is wrong. For example if the value of `csc-owner-label` is set to `installed-foo-openshift-marketplace`, and the `foo-operators` `CatalogSource` does not exist simply because `foo` in `installed-foo-openshift-marketplace` was misplaced in the first place, then the `Subscription` shouldn't reference a non-existent `CatalogSource`. Checking if the `CatalogSource` exists provides us the safety net.

Reply:
Good point regarding B.